toolkit/tools/graphpkgfetcher/graphpkgfetcher.go

```diff
@@ -110,13 +119,85 @@ func fetchPackages() (err error) {
 		return
 	}
 
+	// Optional delta build cache hydration
+	if *tryDownloadDeltaRPMs {
+		logger.Log.Info("Attempting to download delta RPMs for build nodes")
+		err = downloadDeltaNodes(dependencyGraph, cloner)
+		if err != nil {
+			err = fmt.Errorf("failed to download delta RPMs: %w", err)
+			return
+		}
+		// Update the package graph with the paths to the delta RPMs we downloaded
+		err = pkggraph.WriteDOTGraphFile(dependencyGraph, *outputGraph)
+		if err != nil {
+			err = fmt.Errorf("failed to write cache graph to file: %w", err)
+			return
+		}
+	}
+
 	// If we grabbed any RPMs, we need to convert them into a local repo
-	if hasUnresolvedNodes {
+	if *tryDownloadDeltaRPMs || hasUnresolvedNodes {
+		logger.Log.Info("Configuring downloaded RPMs as a local repository")
 		err = cloner.ConvertDownloadedPackagesIntoRepo()
 		if err != nil {
 			err = fmt.Errorf("failed to convert downloaded RPMs into a repo:\n%w", err)
 			return err
 		}
+
+		if strings.TrimSpace(*outputSummaryFile) != "" {
+			err = repoutils.SaveClonedRepoContents(cloner, *outputSummaryFile)
+			if err != nil {
+				logger.Log.Errorf("Failed to save cloned repo contents.")
+				return
+			}
+		}
+	}
+	return
+}
+
+// downloadDeltaNodes will look at the final cached graph we saved and see if any RPMS can be download instead of built.
+// If the previous part of the fetcher worked well we should be able to download only the delta RPMs we need
+// to build our packages or image (i.e. we should be able to create a subgraph just like we would for the build step)
+//   - dependencyGraph: The graph to use to find the packages we need to build. Should have any caching operations already
+//     performed on it. Will be updated with the paths to the delta RPMs we download.
+//   - cloner: The cloner to use to download the RPMs
+func downloadDeltaNodes(dependencyGraph *pkggraph.PkgGraph, cloner *rpmrepocloner.RpmRepoCloner) (err error) {
+	const (
+		deltaBuildGraph = true
+	)
+
+	timestamp.StartEvent("delta package download", nil)
```

Comment:
>timestamp.StartEvent("delta package download", nil) [](http://example.com/codeflow?start=1&length=51)

I've just noticed that `StartEvent` returns an error. We should be checking its value. It may not be a critical failure, if it doesn't start, but we should at least print a warning at that point. Or log it in another way.

Reply:
The idea of the timestamping library is to be non-intrusive and fail-safe. We intentionally chose not to handle errors and just continue. This data is best effort. The errors can be helpful for debugging so they ARE plumbed in but for most use I don't think we want to bother handling them at all. (Maybe to be more explicity we should use `_ = timestamp.StartEvent(...)`?

Reply:
I see and I think I understand the difference between timestamps and the rest of the code. I don't have a better suggestion at the moment, but even if I did, I guess this PR wouldn't be the time to address it.;)